src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+        init,
+
+        source_stat,
+        source_open,
+        source_read,
+        source_swap,
+
+        ready,
+        err: anyerror,
+    } = .init,
+
+    callback: ?Callback = null,
+
+    pub fn init(allocator: std.mem.Allocator, io: *IO, exe_path: [:0]const u8) !Multiversion {
+        assert(builtin.target.os.tag == .linux);
+        assert(std.fs.path.isAbsolute(exe_path));
+
+        // To keep the invariant that whatever has been advertised can be executed, while allowing
+        // new binaries to be put in place, double buffering is used:
+        // * source_buffer is where the in-progress data lives,
+        // * target_fd is where the advertised data lives.
+        // This does impact memory usage.
+        const source_buffer = try allocator.alloc(u8, constants.multiversion_binary_size_max);
+
+        // Only Linux has a nice API for executing from an in-memory file. For macOS and Windows,
+        // a standard named temporary file will be used instead.
+        const target_fd: os.fd_t = switch (builtin.target.os.tag) {
+            .linux => blk: {
+                const fd = open_memory_file(memfd_path);
+                try os.ftruncate(fd, constants.multiversion_binary_size_max);
+
+                break :blk fd;
+            },
+
+            // TODO: macOS / Windows support.
+            else => IO.INVALID_FILE,
+        };
+
+        // We can pass through our env as-is to exec. We have to manipulate the types
+        // here somewhat: they're cast in start.zig and we can't access `argc_argv_ptr`
+        // directly. process.zig does the same trick in execve().
+        //
+        // For args, modify them so that argv[0] is exe_path. This allows our memfd executed binary
+        // to find its way back to the real file on disk.
+        const args = try allocator.allocSentinel(?[*:0]const u8, os.argv.len, null);
+        errdefer allocator.free(args);
+
+        args[0] = try allocator.dupeZ(u8, exe_path);
+        errdefer allocator.free(args[0]);
+
+        for (1..os.argv.len) |i| args[i] = os.argv[i];
+
+        const args_envp = .{
+            .args = args,
+            .envp = @as([*:null]const ?[*:0]const u8, @ptrCast(os.environ.ptr)),
+        };
+
+        return .{
+            .io = io,
+
+            .exe_path = exe_path,
+            .args_envp = args_envp,
+
+            .source_buffer = source_buffer,
+            .target_fd = target_fd,
+        };
+    }
+
+    pub fn start(self: *Multiversion) void {
+        self.read_from_binary(null);
+        self.tick_until_ready() catch {
+            // If there's been an error starting up multiversioning, don't disable it, but
+            // advertise only the current version in memory.
+            self.releases_bundled.clear();
+            self.releases_bundled.append_assume_capacity(constants.config.process.release);
+        };
+
+        log_multiversioning.info("enabling automatic on-disk version detection.", .{});
```

Comment:
Maybe `log.debug()`? I think we want to keep the default output of `tigerbeetle start` quiet, and I don't think this really adds any information.